### ISSUE TITLE
OSD pod goes to sleep when osd daemon is OOMKilled

### ISSUE THREAD

**Issue by sp98** [rook/rook#12972]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
When the OSD daemon is down due to OOMkilled, OSD pod bash scripts reads exit code 0 and makes the OSD pod to sleep for 24 h.
```
[1,0,2] r=2 lpr=20607 pi=[20488,20607)/1 crt=20594'95848 mlcod 0'0 unknown NOTIFY mbc={} ps=[b66~7,b74~7,b82~5,b89~4,b8f~2,b92~1,b96~1,b99~2,b9c~1,b9e~1,ba6~1,baa~1]] state<Start>: transitioning to Stray
debug2023-09-28T11:27:07.822+0000 7f5df5d26640  0 log_channel(cluster) log [DBG] : 11.d scrub starts
debug2023-09-28T11:27:08.026+0000 7f5df5d26640  0 log_channel(cluster) log [DBG] : 11.d scrub ok
debug2023-09-28T11:27:08.871+0000 7f5dfbd32640  0 log_channel(cluster) log [DBG] : 12.1 scrub starts
debug2023-09-28T11:27:20.526+0000 7f5df7d2a640  0 log_channel(cluster) log [DBG] : 12.1 scrub ok
debug2023-09-28T11:27:37.647+0000 7f5dfad30640  0 log_channel(cluster) log [DBG] : 12.13 scrub starts
debug2023-09-28T11:27:47.852+0000 7f5dfad30640  0 log_channel(cluster) log [DBG] : 12.13 scrub ok
debug2023-09-28T11:27:51.828+0000 7f5df952d640  0 log_channel(cluster) log [DBG] : 11.e scrub starts
debug2023-09-28T11:27:52.066+0000 7f5df952d640  0 log_channel(cluster) log [DBG] : 11.e scrub ok
debug2023-09-28T11:28:51.854+0000 7f5df852b640  0 log_channel(cluster) log [DBG] : 5.0 scrub starts
debug2023-09-28T11:28:52.296+0000 7f5df852b640  0 log_channel(cluster) log [DBG] : 5.0 scrub ok
debug2023-09-28T11:28:52.891+0000 7f5df9d2e640  0 log_channel(cluster) log [DBG] : 11.1d scrub starts
debug2023-09-28T11:28:53.130+0000 7f5df5d26640  0 log_channel(cluster) log [DBG] : 11.1d scrub ok
/bin/bash: line 9:   412 Killed                  ceph-osd --foreground --id 2 --fsid 928c8604-6082-4802-893c-3222d5f150ae --setuser ceph --setgroup ceph --crush-location="root=default host=ocs-deviceset-thin-csi-odf-1-data-0457vv rack=rack2" --osd-op-num-threads-per-shard=2 --osd-op-num-shards=8 --osd-recovery-sleep=0 --osd-snap-trim-sleep=0 --osd-delete-sleep=0 --bluestore-min-alloc-size=4096 --bluestore-prefer-deferred-size=0 --bluestore-compression-min-blob-size=8192 --bluestore-compression-max-blob-size=65536 --bluestore-max-blob-size=65536 --bluestore-cache-size=3221225472 --bluestore-throttle-cost-per-io=4000 --bluestore-deferred-batch-ops=16 --default-log-to-stderr=true --default-err-to-stderr=true --default-mon-cluster-log-to-stderr=true --default-log-stderr-prefix=debug --default-log-to-file=false --default-mon-cluster-log-to-file=false --ms-learn-addr-from-peer=false --public-addr=242.0.255.251 --public-bind-addr=10.131.0.41 --cluster-addr=10.131.0.41
+ RESTART_INTERVAL=24
+ rc=0
+ '[' 0 -eq 0 ']'
+ touch /tmp/osd-sleep
OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for 24 hours. Restart the pod manually once the flapping issue is fixed
+ echo 'OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for 24 hours. Restart the pod manually once the flapping issue is fixed'
+ wait
+ sleep 24h
```
**Expected behavior:**
OSD pod should sleep for 24h only in case of OSD flapping.

**How to reproduce it (minimal and precise):**

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary

**Logs to submit**:

* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

  To get logs, use `kubectl -n <namespace> logs <pod name>`
  When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
  Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Cluster Status to submit**:

* Output of kubectl commands, if necessary

  To get the health of the cluster, use `kubectl rook-ceph health`
  To get the status of the cluster, use `kubectl rook-ceph ceph status`
  For more details, see the [Rook kubectl Plugin](https://rook.io/docs/rook/latest-release/Troubleshooting/kubectl-plugin)

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)):

**Comment by travisn** [comment]
Note that this issue does not affect any Rook releases. This is related to #12715 which was backported to the release-1.12 branch, so we will need to fix this before v1.12.5.

**Comment by BlaineEXE** [comment]
I think the main problem here is that `wait` has a return code that is masking the return code of `ceph-osd ...`. Below is the script as-is, with comments.


```sh
function sigterm() {
	echo "SIGTERM received"
        # this still leaves any child processes running, which could zombie
	exit # <-- should exit 143 on sigterm, 137 on SIGKILL
}

trap sigterm SIGTERM # <-- does not capture SIGKILL signal

%s %s & wait # wait should be on the next line after the '&' to fork the process
# more importantly, wait's return code is masking the return code for the main process

RESTART_INTERVAL=%d # this may have a return code that is masking the '%s %s' code (move to beginning)
rc=$?
if [ $rc -eq 0 ]; then
	touch /tmp/osd-sleep
	echo "OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for $RESTART_INTERVAL hours. Restart the pod manually once the flapping issue is fixed"
	sleep "$RESTART_INTERVAL"h & wait # ditto, wait on new line
	exit $rc
fi
```


There are also some other corner cases I see around signal handling that make me nervous. In general, I have always found that it is best to not handle signals in bash scripts or get return codes from forked processes unless absolutely necessary.

In some of my initial testing, this script works fine and handles `kill` and `kill -9` on my machine, though it may be different running as PID 1 in the container. I'll have to look into that.

```sh
RESTART_INTERVAL=%d

sleep 5 && false # <-- OSD command goes here
rc="$?"
if [[ $rc -eq 0 ]]; then
  touch /tmp/osd-sleep
  echo "OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for $RESTART_INTERVAL hours. Restart the pod manually once the flapping issue is fixed"
  sleep "$RESTART_INTERVAL"h
fi
exit $rc
```

**Comment by travisn** [comment]
Testing the below script in the OSD, here are some observations:
- Deleting the OSD pod does not exit immediately. It waits for the 30s pod timeout, as if the SIGTERM is ignored. (not expected)
- If I exec into the OSD pod and run `kill <osdPID>`, the OSD process exits, but the return code appears as 0 and the pod sleeps. (not expected)
- If I exec into the OSD pod and run `kill -9 <osdPID>`, the process exits with code 137. (expected)
```
RESTART_INTERVAL=%d

%s %s

rc=$?
if [ $rc -eq 0 ]; then
	touch /tmp/osd-sleep
	echo "OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for $RESTART_INTERVAL hours. Restart the pod manually once the flapping issue is fixed"
	sleep "$RESTART_INTERVAL"h
fi
exit $rc
```
Notice that we always `exit $rc` instead of just inside the `if` statement.

**Comment by BlaineEXE** [comment]
Bummer. I guess we do need some signal handling, but that means we also need to improve the script to send signals to the child process (ceph-osd or sleep, whichever is running), and also make sure we use working methods to get the return code of the ceph-osd process.

**Comment by BlaineEXE** [comment]
While we are making changes, we should also set the maximum `FlappingRestartIntervalHours` value as 596523, which is `sleep`'s limit in hours.

**Comment by BlaineEXE** [comment]
Based on a good stackexchange I found, this should be a "good enough" script that isn't wildly complicated. https://unix.stackexchange.com/a/444676 mentions a race condition, but hitting that will be extremely rare since it requires pod stop to happen within milliseconds of the script starting. And since Kubernetes will send `kill -9` after 30 seconds, we are protected from the worst effects.

```sh
# see http://veithen.io/2014/11/16/sigterm-propagation.html and
# https://unix.stackexchange.com/questions/146756/forward-sigterm-to-child-in-bash

RESTART_INTERVAL=%d

child_pid=""
function sigterm() {
  echo "SIGTERM received"
  kill -TERM "$child_pid" # no negative effects other than error message if PID not running
}
trap sigterm SIGTERM

%s %s &
# un-fixable race condition: if receive sigterm here, it won't be sent to child process
child_pid="$!"
wait "$child_pid" # wait returns the same return code of child process when called with argument
wait "$child_pid" # first wait returns immediately upon SIGTERM, so wait again for child to actually stop; this is a noop if child exited normally
ceph_osd_rc=$?
if [ $ceph_osd_rc -eq 0 ]; then
  touch /tmp/osd-sleep
  echo "OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for $RESTART_INTERVAL hours. Restart the pod manually once the flapping issue is fixed"
  sleep "$RESTART_INTERVAL"h &
  child_pid="$!"
  wait "$child_pid"
  wait "$child_pid" # wait again for sleep to stop
fi
exit $ceph_osd_rc
```

Be sure to test this thoroughly using Travis' methods [here](https://github.com/rook/rook/issues/12972#issuecomment-1743857023). It's possible I made a mistake with my basic investigation, but the overall script should be fine

**Comment by sp98** [comment]
``` ❯ sh bashtest.sh
+ child_pid=
+ trap sigterm SIGTERM
+ child_pid=44282
+ wait 44282
+ sleep 30
+ wait 44282
bashtest.sh: line 16: wait: pid 44282 is not a child of this shell
+ RESTART_INTERVAL=24
+ child_rc=0
+ '[' 0 -eq 0 ']'
+ touch /tmp/osd-sleep
+ echo 'OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for 24 hours. Restart the pod manually once the flapping issue is fixed'
OSD daemon exited with code 0, possibly due to OSD flapping. The OSD pod will sleep for 24 hours. Restart the pod manually once the flapping issue is fixed
+ child_pid=44369
+ wait 44369
+ sleep 24h
```

does not really work when I kill the first sleep of 30 seconds.

**Comment by travisn** [comment]
Not blocking 1.12 anymore since the revert was merged in #13005. Following up in master with #12992.
